src/search.rs

```diff
@@ -24,6 +25,53 @@ pub trait SearchIndex: BackwardIterableIndex {
     {
         Search::new(self).search(pattern)
     }
+
+    // If we created a HasDoc trait (or something better named) for those
```

Comment:
It might be also an option that we implement new APIs on a new trait/struct such as `MultiTextSearchIndex` that _wraps_ a `BackwardIterableIndex` instance.

- We can avoid adding `Doc` data to every FM-Index variant like `FMIndex` and `RLFMIndex`.
- We can consider different APIs for single-text search index `SearchIndex` and multi-texts search index `MultiTextSearchIndex`. For instance, `locate` query in the former trait may just return pattern positions, whereas the latter may return positions with `TextId`.

Reply:
Yes, I think it would be nice to have a search index API that doesn't allow any \0 (or at least one at the end), and a multi index that accepts \0 (or a builder pattern), that offers the extended APIs.

So let's explore this. I'm a still not entirely sure how to implement doc so I hope to get this in a shape so you can help me implement it.

Reply:
Got it! I'll also try to implement the multi-text search based on the APIs you outlined.